### PR TITLE
Organizations: Test access from anonymous users and users outside the organization

### PR DESCRIPTION
**Pull request by stsewd**
Currently, I'm checking the global option
of `ALLOW_PRIVATE_REPOS` to give access to public views
to users outside the org.
I'm using the same patter we already have with projects about having a `public` queryset.

We could introduce a privacy level for the organization if we want.

### CODE REVIEW

**Comment by ericholscher** [comment]
@stsewd is this still a valid PR? Seems useful, so we should review it if so.

**Comment by stsewd** [comment]
> @stsewd is this still a valid PR? Seems useful, so we should review it if so.

It's, it should be ready now

File: readthedocs/organizations/managers.py [hunk 1]
```diff
@@ -30,6 +35,17 @@ def teams_for_user(self, user, organization, admin, member):
 
         return teams.distinct()
 
+    def public(self, user):
+        """
+        Return all teams the user has access to.
+
+        If ``ALLOW_PRIVATE_REPOS`` is `False`, all teams are public by default.
+        Otherwise, we return only the teams where the user is a member.
```
Comment (ericholscher) [inline readthedocs/organizations/managers.py:43]: Huh.. do we want teams to always be public? I guess the example here is Orgs & Teams on .org once we enable them. I guess I'm fine defaulting them public, since everything is default public now.

File: readthedocs/organizations/querysets.py [hunk 2]
```diff
@@ -15,13 +15,29 @@ class BaseOrganizationQuerySet(models.QuerySet):
...
+        if not user.is_authenticated:
+            return self.none()
         return self.filter(
             Q(owners__in=[user]) | Q(teams__members__in=[user]),
         ).distinct()
 
     def for_admin_user(self, user):
+        """List all organizations where the user is an owner."""
+        if not user.is_authenticated:
+            return self.none()
```
Comment (ericholscher) [inline readthedocs/organizations/querysets.py:40]: I assume these are just optimizations?

**Review by ericholscher — COMMENTED** [review]
This looks like a solid change. Also asking @humitos for a review to ✅  it, since I'm not 100% the implications here. 

> We could introduce a privacy level for the organization if we want.

Privacy levels are already quite confusing. We definitely don't want to add more of them unless we really need to.

File: readthedocs/organizations/managers.py (hunk 1 above)
Comment (humitos) [reply readthedocs/organizations/managers.py:43]: I had the same thought as Eric here. Where, or in which situations, are we going to show these teams to anonymous users?

How do other platforms work here? In GitHub for example, I don't think you have access to see all the teams inside an organization. For example, in ours https://github.com/orgs/readthedocs/people, I don't think we have a way to list `backend`, `frontend` and `advocate` teams there. However, on PRs I can see those teams publicly: https://github.com/readthedocs/readthedocs.org/pull/929. When logged in, I can access to its members as well https://github.com/orgs/readthedocs/teams/backend

Another thing that GitHub has related to permissions on teams/organizations is "I want to show publicly that I'm member of this organization". Which adds another layer of complexity here.

I'm just describing how another platform works, I'm not saying we need to implement any of this, but just to keep in mind where are we going with this and how do we want it to work.

File: readthedocs/organizations/tests/test_privacy_urls.py [hunk 3]
```diff
@@ -74,3 +76,195 @@ def test_public_urls(self):
...
+        },
+    }
+
+    def login(self):
+        self.client.force_login(self.another_user)
+
+    def test_private_urls(self):
+        from readthedocs.organizations.urls.private import urlpatterns
+
+        self._test_url(urlpatterns)
```
Comment (humitos) [inline readthedocs/organizations/tests/test_privacy_urls.py:270]: Pretty hard to follow all these tests with such magic. None of them seems to do anything different one from the other (just a setting and URLs). I assume that _all this magic_ is inside this method :/

I assume this is fine, tho, to avoid repetition, but... :(

File: readthedocs/organizations/urls/public.py [hunk 4]
```diff
@@ -4,11 +4,6 @@
 from readthedocs.organizations.views import public as views
 
 urlpatterns = [
-    url(
-        r'^verify-email/$',
-        views.OrganizationTemplateView.as_view(template_name='organizations/verify_email.html'),
-        name='organization_verify_email',
-    ),
```
[the remark this answers is not in the document]
Comment (humitos) [reply readthedocs/organizations/urls/public.py:11]: Makes sense!

File: readthedocs/organizations/querysets.py [hunk 5]
```diff
@@ -15,13 +15,29 @@ class BaseOrganizationQuerySet(models.QuerySet):
 
     """Organizations queryset."""
 
+    def public(self, user):
+        """
+        Return all organizations the user has access to.
+
+        If ``ALLOW_PRIVATE_REPOS`` is `False`, all organizations are public by default.
```
Comment (humitos) [inline readthedocs/organizations/querysets.py:22]: I think we should probably rename this to `RTD_CORPORATE` or something. We have been using this setting to differentiate if we are on .org or .com lately, but it's not what the name suggests.

File: readthedocs/organizations/views/base.py [hunk 6]
```diff
@@ -72,7 +72,7 @@ def get_organization_queryset(self):
         """
         if self.admin_only:
             return Organization.objects.for_admin_user(user=self.request.user)
-        return Organization.objects.for_user(user=self.request.user)
+        return Organization.objects.public(user=self.request.user)
```
Comment (humitos) [inline readthedocs/organizations/views/base.py:75]: We don't have the concept of privacy level in Organizations, so I'm not sure these changes make a ton of sense.

File: readthedocs/organizations/views/base.py [hunk 7]
```diff
@@ -109,7 +109,11 @@ def get_team_queryset(self):
         This will either be team the user is a member of, or teams where the
         user is an owner of the organization.
         """
-        return Team.objects.member(self.request.user).filter(
+        if self.admin_only:
```
Comment (humitos) [inline readthedocs/organizations/views/base.py:112]: Where are we setting this `self.admin_only`?

**Review by humitos — COMMENTED** [review]
I took a look at this and the changes look fine. However, I'm a little worried about changing the auth logic without understanding exactly why 😄 . Basically, why these changes are needed? What are their benefits? I'm not sure to follow the goal of the PR. Is it just a refactor?

We are now using `.public` instead of `.for_user`, but there is nothing extra returned on `.public` methods since we don't have privacy levels in organizations, so 🤷🏼 . I'm confused...

**Comment by ericholscher** [comment]
@stsewd would be good to address the review comments here and get this merged, if possible.

**Comment by agjohnson** [comment]
Ping @stsewd :point_up:

We can also push this out some sprints or back on the backlog if it makes sense to return to this later instead.

**Comment by ericholscher** [comment]
Pushed it back a few sprints 👍

**Comment by humitos** [comment]
I'm putting this back into Q2, so we come back to this PR soon.

**Comment by agjohnson** [comment]
Doing some pruning here. How likely are we to get back to this PR? Is it worth prioritizing in the near future, or should we discuss retiring this work?

**Comment by stsewd** [comment]
> Doing some pruning here. How likely are we to get back to this PR? Is it worth prioritizing in the near future, or should we discuss retiring this work?

I'm fine closing this, it may be helpful when we decide to introduce organizations in community.
